dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/css/OpenAPIUIBundlesUpdater.java

```diff
@@ -313,28 +302,17 @@ private static void writeStreamToZip(InputStream stream, ZipOutputStream zos) th
     @SuppressWarnings("unchecked")
     private static boolean waitForBundlesToStart(Set<Bundle> openAPIUIBundles) {
         boolean waitComplete = false;
+
         Set<String> expectedBundleNames = new HashSet<>();
         for (Bundle bundle : openAPIUIBundles) {
             expectedBundleNames.add(bundle.getSymbolicName());
         }
 
         try {
             BundleContext bundleContext = FrameworkUtil.getBundle(OpenAPIUIBundlesUpdater.class).getBundleContext();
-            ServiceReference<Bundle>[] refs = (ServiceReference<Bundle>[]) bundleContext.getServiceReferences(Bundle.class.getName(), "(installed.wab.contextRoot=*)");
-            waitComplete = verifyRefs(refs, expectedBundleNames);
-            while (!waitComplete) {
-                if (stopping) {
-                    if (LoggingUtils.isDebugEnabled(tc)) {
-                        Tr.debug(tc, "Stopped waiting for OpenAPI bundles because the server is shutting down");
-                    }
-                    return false;
-                }
-                refs = (ServiceReference<Bundle>[]) bundleContext.getServiceReferences(Bundle.class.getName(), "(installed.wab.contextRoot=*)");
-                waitComplete = verifyRefs(refs, expectedBundleNames);
-                if (!waitComplete) {
-                    //sleep UI bundle update thread to prevent getting services refs all the time
-                    Thread.sleep(500);
-                }
+            bundleContext.addServiceListener(new OpenAPIUIBundlesListener(expectedBundleNames));
+            synchronized (threadLock) {
+                threadLock.wait();
```

Comment:
I suggest you use a CountDownLatch here and initialize it with the size of the expectedBundleNames.  Then pass that latch to the constructor of `OpenAPIUIBundlesListener`.  That way the listener can `countDown()` on the latch each time an expected bundle is found. and you can `await()` for it here.

There is also an issue with discovery of existing service registrations that match.  The listener is only going to see new events, it will not be able to detect any existing services here.  I suggest you add a method to the listener class (called `checkExistingServices` or similar) and that gets the existing services to check:

```
ServiceReference<Bundle>[] existingRefs = (ServiceReference<Bundle>[]) bundleContext.getServiceReferences(Bundle.class.getName(), "(installed.wab.contextRoot=*)");
```
And it checks reach existing ref and also does a `countDown()` when they are for one of the expected bundles.